lib/fast_jsonapi/relationship.rb

```diff
@@ -120,5 +153,74 @@ def run_key_transform(input)
         input.to_sym
       end
     end
+
+    def initialize_static_serializer
```

Comment:
@christophersansone I'm trying to understand the reasoning behind all this caching and extra attributes. Correct me if I'm wrong, but a relationship will have just one serializer klass, which we can calculate upon the relationship initialization DSL (say as part of the `create_relationship`). Same is true for the polymorphic case. So why do we need the extra logic here, seems overly complicated?

Reply:
> Correct me if I'm wrong, but a relationship will have just one serializer klass...

This is true in some cases but not all.  A relationship's serializer can either be _static_ or _dynamic_.

When it is static, it is determined once up front and will not change.  This is the most common use case:

```ruby
belongs_to :actor
has_many :movies
belongs_to :genre, serializer: CustomGenreSerializer
```

The serializer is considered _dynamic_ when it may change based on `object` or `params`.  This occurs in the following ways:

```
has_many :movies, polymorphic: true
has_many :actors, polymorphic: { ComedyActor => :custom_comedy, DramaActor => :custom_drama }
has_many :genres, serializer: -> (object, params) {
  ...
}
```

Therefore, the role of `initialize_static_serializer` is to determine _if_ the relationship's serializer can be static, and if so, which serializer it is.  It determines this once and then caches it forever.  If it determines that the serializer is dynamic, then it returns `nil`, stating that there is no static serializer.

Then, if there is no static serializer, it knows that it needs to dynamically determine the serializer.  In some cases, like `polymorphic: true`, it simply looks at the object class, computes the serializer by name once and memoizes it in `@serializers_for_name`.

The bottom line is this.  If `options[:serializer]` is specified, that takes precedence over everything.  The developer is specifically stating which serializer to use (or how to determine it if a Proc is specified).  If not, then it needs to infer it, based on whether `options[:polymorphic]` is specified, by the relationship name, etc.

Is it complicated?  Yes.  Is it overly complicated?  I don't think so, given that it all needed to be backwards compatible with `options[:record_type]` and `options[:polymorphic]`.  I basically just consolidated all the logic into one place.  Now to me the flow is quite simple:

1. Up front, it determines if the serializer is static or dynamic.
2. For every record being serialized, the relationship calls `serializer_for` to determine which serializer to use.  If the serializer is static, it returns the static serializer immediately.  If not, it performs the dynamic check (which also does what it can to memoize the results for performance purposes).

To get an idea of how much more consolidated the logic is, look in the original implementation about how many checks for `polymorphic` are all over the place.  Look at how many times it needs to iteratively call `serializer.to_s.constantize` or `object.class.name.demodulize.underscore.to_sym.to_s.constantize`.  And consider how much simpler the `Relationship` API is for determining the serializer of any given object.

Finally, to address the second part of your question:

> ...which we can calculate upon the relationship initialization DSL (say as part of the create_relationship)...

If the serializer is indeed static, then it is indeed determined "up front".  It is not done in `create_relationship` for two reasons:

1. As stated earlier, if `options[:serializer]` is specified, then it takes precedence over any other serializer determination such as `polymorphic`.  Therefore, we don't want to create the relationship with its serializer defined as `options[:serializer] || compute_serializer()`.  We _could_ still compute it in `create_relationship`, but it would have to be as a separate variable.  To me this is more complicated, not less, and at that point I would rather have the relationship handle it.

2. Before, the `serializer` passed into `Relationship` was a symbol, which still has to be constantized.  And it was constantized often.  Now, the static serializer determined by the relationship is the actual serializer class, and THAT is memoized.  Because of this, we cannot necessarily determine this when the Relationship is created, because the serializer class may not be loaded yet.  Therefore, it is determined "up front", i.e. the first time it is attempted.

Hopefully this answers all your questions.